enclave/impl/src/message_bus/marshalled_record.rs

```diff
@@ -0,0 +1,281 @@
+/// marshalled to the layout of a value in the oblivious message bus.
+///
+/// The only intended implementer is A8Bytes<MessageBusValueSize>.
+///
+/// The layout is:
+/// 0..32 = sender bytes
+/// 32..64 = recipient bytes
+/// 64..72 = timestamp bytes (native endian)
+/// 72...  = payload bytes
+///
+/// The total size is determined by the MessageBusValueSize constant, which is
+/// determined by the choice of ORAM parameters, and the size of the message bus
+/// key.
+///
+/// The payload, and record, are always fixed size given these parameters.
+pub trait MarshalledRecord: Sized {
+    /// Create a MarshalledRecord from a RequestRecord proto object
+    ///
+    /// Returns:
+    /// * A MsgId object, and Self
+    /// * Errors if any of the fields in the protobuf are the wrong size
+    fn new_from_proto(proto_record: &RequestRecord) -> Result<(MsgId, Self), QueryError>;
+
+    /// Take a MarshalledRecord, and the associated message bus key, and form a
+    /// protobuf Record object, unmarshalling the data and formatting it for
+    /// return to the user.
+    fn to_proto(&self, msg_id: &MsgId) -> Record;
+
+    /// Get a reference to the bytes which correspond to the sender.
+    fn get_sender(&self) -> &A8Bytes<U32>;
+
+    /// Get a mutable reference to the bytes which correspond to the sender.
+    fn get_sender_mut(&mut self) -> &mut A8Bytes<U32>;
+
+    /// Get a reference to the bytes which correspond to the recipient.
+    fn get_recipient(&self) -> &A8Bytes<U32>;
+
+    /// Get a mutable reference to the bytes which correspond to the recipient.
+    fn get_recipient_mut(&mut self) -> &mut A8Bytes<U32>;
+
+    /// Get a reference to the timestamp, as a u64
+    fn get_timestamp(&self) -> &u64;
+
+    /// Get a mutable reference to the timestamp, as a u64
+    fn get_timestamp_mut(&mut self) -> &mut u64;
+
+    /// Get a reference to the payload bytes
+    fn get_payload(&self) -> &A8Bytes<Diff<MessageBusValueSize, U72>>;
+
+    /// Get a mutable reference to the payload bytes
+    fn get_payload_mut(&mut self) -> &mut A8Bytes<Diff<MessageBusValueSize, U72>>;
+}
+
+impl MarshalledRecord for A8Bytes<MessageBusValueSize> {
+    fn new_from_proto(proto_record: &RequestRecord) -> Result<(MsgId, Self), QueryError> {
+        let mut msg_id = MsgId::default();
+        if proto_record.msg_id.len() != msg_id.as_slice().len() {
+            return Err(QueryError::InvalidMessageIdLength(
+                proto_record.msg_id.len(),
+                msg_id.as_slice().len(),
+            ));
+        }
+        msg_id
+            .as_mut_slice()
+            .copy_from_slice(&proto_record.msg_id[..]);
+
+        let mut data: A8Bytes<MessageBusValueSize> = Default::default();
+        let (_sender, rest): (&mut A8Bytes<U32>, &mut _) = (&mut data).split();
+        let (recipient, rest): (&mut A8Bytes<U32>, &mut _) = rest.split();
+        let (_timestamp, rest): (&mut A8Bytes<U8>, &mut _) = rest.split();
```

Comment:
Consider renaming `rest` to `payload` since thats what it is after this split.

Reply:
Since I see more splits down below, might be worth making a helper method in this file that gets a A8Bytes<MessageBusValueSize> and returns the 4 parts

Reply:
maybe i can just use `get_recipient_mut()` at this point

i might still try to add a helper function though